### PR TITLE
Avoid dangling pointers on unit tests

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
This change corrects all places where dangling pointers were observed to cause crashes when running `brave_unit_tests`. Most fixes are merely a matter of changing the initialisation order of members, with others requiring `raw_ptr` instances to be cleared out when appropriate.

Resolves https://github.com/brave/brave-browser/issues/33265

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by zenparsing — APPROVED** [review]
Rewards 👍

File: components/brave_rewards/browser/rewards_service_impl_unittest.cc [hunk 1]
```diff
@@ -78,8 +79,8 @@ class RewardsServiceTest : public testing::Test {
     auto* rewards_ = new RewardsServiceImpl(profile());
 #endif
     RewardsServiceFactory::SetServiceForTesting(std::move(rewards_));
-    rewards_service_ = static_cast<RewardsServiceImpl*>(
-        RewardsServiceFactory::GetForProfile(profile()));
+    rewards_service_ = base::WrapUnique(static_cast<RewardsServiceImpl*>(
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl_unittest.cc:82]: This whole thing seems strange. What's the point of getting `rewards_service_` from the factory when we set `rewards_` as the service to be returned by the factory? It seems like `SetServiceForTesting` should just go away entirely and either use the directly created `rewards_service_` or use the one returned by `RewardsServiceFactory::GetForProfile` and not do https://github.com/brave/brave-core/pull/20306/files#diff-79a8ff13d4400b7fb9810d76aa14728acb0aded37d2dc1c414a8b00396489642R77 at all

File: components/brave_rewards/browser/rewards_service_impl_jp_unittest.cc [hunk 2]
```diff
@@ -51,8 +52,8 @@ class RewardsServiceJPTest : public testing::Test {
     auto* rewards_ = new RewardsServiceImpl(profile());
 #endif
     RewardsServiceFactory::SetServiceForTesting(std::move(rewards_));
-    rewards_service_ = static_cast<RewardsServiceImpl*>(
-        RewardsServiceFactory::GetForProfile(profile()));
+    rewards_service_ = base::WrapUnique(static_cast<RewardsServiceImpl*>(
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl_jp_unittest.cc:55]: see below

**Review by DJAndries — APPROVED** [review]
p3a++

File: components/brave_rewards/browser/rewards_service_impl_unittest.cc (hunk 1 above)
Comment (cdesouza-chromium) [reply components/brave_rewards/browser/rewards_service_impl_unittest.cc]: Fixed a21f6b7c5d5353cc933ca9ecf3eac77e26d727ab.

File: components/brave_rewards/browser/rewards_service_impl_jp_unittest.cc (hunk 2 above)
Comment (cdesouza-chromium) [reply components/brave_rewards/browser/rewards_service_impl_jp_unittest.cc]: Fixed a21f6b7c5d5353cc933ca9ecf3eac77e26d727ab.

**Review by bridiver — APPROVED** [review]

**Review by yrliou — APPROVED** [review]
++wallet, ddns, ipfs

**Review by iefremov — APPROVED** [review]

**Pull request merged**